content/blog/2021-12-07-ml-experiment-versioning.md

```diff
@@ -0,0 +1,240 @@
+they may store binary artifacts like input data or model weights so that you can
+reproduce experiments and retrieve results. They enable you to navigate all this
+meta-information, for example with dashboards to organize and compare many
+experiments.
+
+Git can't manage or compare all that experiment meta-information, but it is
+still better for code. For example, Git tracks incremental changes instead of
+having to log all code files as artifacts for each experiment. Therefore,
+experiments get split between Git for code and experiment tracking tools for
+meta-information. A link may be added in one or the other to keep track.
+
+ML experiment versioning combines experiment tracking and version control:
+
+- **Experiments as code**: Track meta-information in the code repository and
+  version it like code.
+- **Versioned reproducibility**: Save and restore experiment state, and track
+  changes to only execute what's new.
+- **Distributed experiments**: Organize locally and choose what to share,
+  reusing your existing repo setup.
+
+```
++---------------------+    +------------------+
+|                     |    |                  |
+| Experiment Tracking |    |  Code Versioning |
+|                     |    |                  |
++---------------------+    +---------/--------+
+            \                       /
+             \                     /
+              \-                 /-
+                \               /
+                 \             /
+            +-----\------------------+
+            |                        |
+            | Experiment Versioning  |
+            |                        |
+            +------------------------+
+```
+
+# ML Experiments as Code
+
+Experiment versioning treats experiments as code. Experiment tracking tools log
+parameters, metrics, and artifacts to a database through API calls. In
+experiment versioning, they are saved in files.
+
+You can choose your own file formats and paths, which you can configure in DVC:
+
+```dvc
+$ dvc exp init -i
+This command will guide you to set up a default stage in dvc.yaml.
+See https://dvc.org/doc/user-guide/project-structure/pipelines-files.
+
+DVC assumes the following workspace structure:
+├── data
+├── metrics.json
+├── models
+├── params.yaml
+├── plots
+└── src
+
+Command to execute: python src/train.py
+Path to a code file/directory [src, n to omit]: src/train.py
+Path to a data file/directory [data, n to omit]: data/images/
+Path to a model file/directory [models, n to omit]:
+Path to a parameters file [params.yaml, n to omit]:
+Path to a metrics file [metrics.json, n to omit]:
+Path to a plots file/directory [plots, n to omit]: logs.csv
+```
+
+This might mean a little more work to set up your code to read parameters from a
+file or write metrics to another file. To track artifacts like data and models
+in your repo, DVC adds
+[data versioning](https://dvc.org/doc/start/data-and-model-versioning) support.
```

Comment:
Is it really more work compared to exp tracking. It is (conceptually 🙂 ) the same. It feels like this paragraph can be removed.